### ISSUE TITLE
Simple model training failure: loss does not descent.

### ISSUE THREAD

**Issue by deahhh** [pytorch/pytorch#63626]
## 🐛 Bug

Simple model training failure. I am sure the normal loss should be less than 0.01, but here never descent to 0.3 or less.

## To Reproduce

1. make env of python3.8.10 in ubuntu 20.04. Env as follows:
```bash
Python 3.8.10 (default, Jun  2 2021, 10:49:15) 
[GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> torch.__version__
'1.7.1+cu110'
>>> 
```
2. run following code, you will get loss never descent to 0.3 or less.
```python
def test_torch():
    import torch
    import torch.nn as nn
    from torch.utils import data
    from torch.utils.data import dataloader

    data = torch.randn(100000, 2)
    y = torch.sin(data[:, 0]*data[:, 1]).unsqueeze(1)
    print(y.shape)
    data = torch.cat([data, y], dim=1)

    model = nn.Sequential(
        nn.Linear(2, 10),
        nn.ReLU(),
        nn.Linear(10, 10), 
        nn.ReLU(),
        nn.Linear(10, 10), 
        nn.ReLU(),
        nn.Linear(10, 10), 
        nn.ReLU(),
        nn.Linear(10, 10), 
        nn.ReLU(),
        nn.Linear(10, 10), 
        nn.ReLU(),
        nn.Linear(10, 10), 
        nn.ReLU(),
        nn.Linear(10, 1)
    )

    for p in model.parameters():
        print(p.shape, p.requires_grad)

    criterion = torch.nn.L1Loss()
    opt = torch.optim.SGD(model.parameters(), 0.001)

    dl = dataloader.DataLoader(data, batch_size=500)
    for epoch in range(100):
        for d in dl:
            pred = model(d[:, :2])
            loss = criterion(pred, d[:, 2])
            opt.zero_grad()
            loss.backward()
            opt.step()
            print("epoch:[{}], loss: {}".format(epoch, loss.item()))
if __name__ == "__main__":
    test_torch()
```

**Comment by Naaru-01** [comment]
Hello there, I would like to work on this Issue. Could you help to understand it better so that I can Solve it asap ?

**Comment by yogeshchandrasekharuni** [comment]
can you explain how you're sure that the loss should reach 0.01 and that this is not the expected behavior?

**Comment by ejguan** [comment]
Hi,
We use github issues only for bugs or feature requests.
Please use the forum to ask questions:  https://discuss.pytorch.org/

Have you ever tried Adam as the optimizer?

**Comment by deahhh** [comment]
@yogeshchandrasekharuni 
I've tried it in tensorflow version.

**Comment by deahhh** [comment]
@ejguan 
It is actually almost a bug except you device pytorch to convergence better by sacrificing some correctness.

**Comment by ejguan** [comment]
I don't think it's a bug. Your loss is calculated incorrectly. Please check the shape `pred` and `d[:, 2]`.

**Comment by deahhh** [comment]
@ejguan
Thanks!
Unbelievable that such a short piece of code has such an obvious error in it! 
Unfortunately, the fix still doesn't yield convergence.

**Comment by deahhh** [comment]
Should I delete it, #63626?
And how?

**Comment by ejguan** [comment]
It's fine as long as it's closed.
